lib/stat_tracker.rb

```diff
@@ -14,34 +21,309 @@ def self.from_csv(data_files)
+          losses[season] += 1.0
+        elsif game.outcome == :home_win
+          losses[season] -= 1.0
+        else
+          losses[season] += 0.5
+        end
+      end
+    end
+    losses
+  end
+
+  def season_away_losses(team_id)
+    losses = Hash.new(0.0)
+    away_games_grouped_by_season(team_id).each do |season, game|
+      game.each do |game|
+        if game.outcome == :home_win
+          losses[season] += 1.0
+        elsif game.outcome == :away_win
+          losses[season] -= 1.0
+        else
+          losses[season] += 0.5
+        end
+      end
+    end
+    losses
+  end
+
+  def loss_count_by_season(team_id)
+    accum = {}
+    season_away_losses(team_id).each do |season_away_id, away_loss|
+      season_home_losses(team_id).each do |season_home_id, home_loss|
+        if season_away_id == season_home_id
+          accum[season_away_id] = away_loss + home_loss
+        end
+      end
+    end
+    accum
+  end
+
+  def worst_season(team_id)
+    loss_count_by_season(team_id).max_by do |game_season, losses|
+      losses
+    end[0]
+  end
+
+  def combine_all_games_played(team_id)
+    home_games_filtered_by_team(team_id).push(away_games_filtered_by_team(team_id)).flatten
+  end
+
+  def find_total_wins_or_ties(team_id)
+    wins_and_ties = 0.0
+    total_games = combine_all_games_played(team_id).each do |game|
+      if team_id == game.home_team_id && game.outcome == :home_win
+        wins_and_ties += 1.0
+      elsif team_id == game.away_team_id && game.outcome == :away_win
+        wins_and_ties += 1.0
+      end
+    end
+    (wins_and_ties / total_games.count).round(2)
+  end
+
+  def average_win_percentage(team_id)
+    find_total_wins_or_ties(team_id)
+  end
+
+  def most_home_goals_scored(team_id)
+    home_games_filtered_by_team(team_id).max_by do |game|
+      game.home_goals
+    end.home_goals.to_i
+  end
+
+  def most_away_goals_scored(team_id)
+    away_games_filtered_by_team(team_id).max_by do |game|
+      game.away_goals
+    end.away_goals.to_i
+  end
+
+  def most_goals_scored(team_id)
+    if most_home_goals_scored(team_id) > most_away_goals_scored(team_id)
+      most_home_goals_scored(team_id)
+    else
+      most_away_goals_scored(team_id)
+    end
+  end
+
+  def fewest_home_goals_scored(team_id)
+    home_games_filtered_by_team(team_id).min_by do |game|
+      game.home_goals
+    end.away_goals.to_i
+  end
+
+  def fewest_away_goals_scored(team_id)
+    away_games_filtered_by_team(team_id).min_by do |game|
+      game.away_goals
+    end.away_goals.to_i
+  end
+
+  def fewest_goals_scored(team_id)
+    if fewest_home_goals_scored(team_id) < fewest_away_goals_scored(team_id)
+      fewest_home_goals_scored(team_id)
+    else
+      fewest_away_goals_scored(team_id)
+    end
+  end
+
+  def all_games_played_by_team(team_id)
+    game_collection.all.find_all do |game|
+      game.home_team_id == team_id || game.away_team_id == team_id
+    end
+  end
+
+  def opponents(team_id)
```

Comment:
This method is quite long and can definitely be refactored to potentially use `reduce` instead of `each`.